File: packages/core/src/research_engine/domain/provenance.py
"""Provenance and operations domain types."""

from __future__ import annotations

import enum
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from research_engine.domain.common import IngestionItemStatus, IngestionRunStatus


class LLMCall(BaseModel):
    """A logged LLM call for auditability and cost tracking."""

    id: UUID
    purpose: str
    caller: str
    model: str
    input_tokens: int | None = None
    output_tokens: int | None = None
    cost_estimate: float | None = None
    duration_ms: int | None = None
    status: str
    error: str | None = None
    created_at: datetime


class LLMCallDraft(BaseModel):
    """Data needed to log an LLM call."""

    purpose: str
    caller: str
    model: str
    input_tokens: int | None = None
    output_tokens: int | None = None
    cost_estimate: float | None = None
    duration_ms: int | None = None
    status: str
    error: str | None = None


class UsageGroup(BaseModel):
    """Spend and token counts for one combination of grouping keys."""

    key: dict[str, str]
    calls: int
    input_tokens: int
    output_tokens: int
    cost: float
    errors: int


class UsageSummary(BaseModel):
    """Aggregated LLM spend over a time window."""

    since: datetime | None = None
    until: datetime | None = None
    group_by: list[str]
    groups: list[UsageGroup]
    total_calls: int
    total_cost: float


class BudgetExceeded(Exception):
    """Configured LLM spend limit reached; the call was refused, not attempted."""

    def __init__(self, spent: float, limit: float, window_days: int) -> None:
        self.spent = spent
        self.limit = limit
        self.window_days = window_days
        super().__init__(
            f"LLM budget exceeded: ${spent:.2f} spent in the last {window_days}d "
            f"against a ${limit:.2f} limit. Raise RE_LLM_BUDGET_USD or wait for "
            f"the window to roll over."
        )


class IngestionRun(BaseModel):
    """A batch ingestion run record."""

    id: UUID
    started_at: datetime
    completed_at: datetime | None = None
    source_spec: dict[str, Any] = Field(default_factory=dict)
    status: IngestionRunStatus
    stats: dict[str, Any] = Field(default_factory=dict)


class IngestionItem(BaseModel):
    """A single item within an ingestion run."""

    id: UUID
    run_id: UUID
    source_ref: str
    document_id: UUID | None = None
    status: IngestionItemStatus
    error: str | None = None
    duration_ms: int | None = None
    created_at: datetime


class PluginActivationState(enum.StrEnum):
    available = "available"
    enabled = "enabled"
    disabled = "disabled"
    pending_approval = "pending_approval"
    missing = "missing"
    incompatible = "incompatible"
    error = "error"
    legacy = "legacy"


class PluginActivation(BaseModel):
    """Approval and runtime state for one installed plugin distribution."""

    plugin_id: str
    distribution_name: str | None = None
    distribution_version: str
    entry_point_name: str | None = None
    manifest_sha256: str | None = None
    manifest: dict[str, Any]
    permissions_granted: dict[str, Any]
    installed_at: datetime
    enabled: bool = False
    state: PluginActivationState = PluginActivationState.legacy
    approved_at: datetime | None = None
    approved_non_interactive: bool = False
    last_seen_at: datetime | None = None
    last_error: str | None = None
    provenance: dict[str, Any] | None = None
    legacy_source_url: str | None = None
    legacy_source_ref: str | None = None
    database_revision: int | None = None
    database_status: str | None = None

    @model_validator(mode="after")
    def _approved_rows_have_distribution_identity(self) -> PluginActivation:
        if self.state is PluginActivationState.legacy:
            return self
        required = {
            "distribution_name": self.distribution_name,
            "entry_point_name": self.entry_point_name,
            "manifest_sha256": self.manifest_sha256,
            "approved_at": self.approved_at,
        }
        missing = [name for name, value in required.items() if value is None]
        if missing:
            raise ValueError(
                "non-legacy plugin activation requires " + ", ".join(missing)
            )
        return self

File: crates/marginalia-types/src/provenance.rs
//! Provenance and operations types mirroring `domain/provenance.py`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::common::{IngestionItemStatus, IngestionRunStatus};
use crate::errors::{Error, Result};

/// A logged LLM call for auditability and cost tracking.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LlmCall {
    pub id: Uuid,
    pub purpose: String,
    pub caller: String,
    pub model: String,
    #[serde(default)]
    pub input_tokens: Option<i64>,
    #[serde(default)]
    pub output_tokens: Option<i64>,
    #[serde(default)]
    pub cost_estimate: Option<f64>,
    #[serde(default)]
    pub duration_ms: Option<i64>,
    pub status: String,
    #[serde(default)]
    pub error: Option<String>,
    pub created_at: DateTime<Utc>,
}

/// Data needed to log an LLM call.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LlmCallDraft {
    pub purpose: String,
    pub caller: String,
    pub model: String,
    #[serde(default)]
    pub input_tokens: Option<i64>,
    #[serde(default)]
    pub output_tokens: Option<i64>,
    #[serde(default)]
    pub cost_estimate: Option<f64>,
    #[serde(default)]
    pub duration_ms: Option<i64>,
    pub status: String,
    #[serde(default)]
    pub error: Option<String>,
}

/// Spend and token counts for one combination of grouping keys.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UsageGroup {
    pub key: std::collections::BTreeMap<String, String>,
    pub calls: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cost: f64,
    pub errors: i64,
}

/// Aggregated LLM spend over a time window.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UsageSummary {
    #[serde(default)]
    pub since: Option<DateTime<Utc>>,
    #[serde(default)]
    pub until: Option<DateTime<Utc>>,
    pub group_by: Vec<String>,
    pub groups: Vec<UsageGroup>,
    pub total_calls: i64,
    pub total_cost: f64,
}

/// A batch ingestion run record.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct IngestionRun {
    pub id: Uuid,
    pub started_at: DateTime<Utc>,
    #[serde(default)]
    pub completed_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub source_spec: Map<String, Value>,
    pub status: IngestionRunStatus,
    #[serde(default)]
    pub stats: Map<String, Value>,
}

/// A single item within an ingestion run.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct IngestionItem {
    pub id: Uuid,
    pub run_id: Uuid,
    pub source_ref: String,
    #[serde(default)]
    pub document_id: Option<Uuid>,
    pub status: IngestionItemStatus,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub duration_ms: Option<i64>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PluginActivationState {
    Available,
    Enabled,
    Disabled,
    PendingApproval,
    Missing,
    Incompatible,
    Error,
    #[default]
    Legacy,
}

/// Approval and runtime state for one installed plugin distribution.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PluginActivation {
    pub plugin_id: String,
    #[serde(default)]
    pub distribution_name: Option<String>,
    pub distribution_version: String,
    #[serde(default)]
    pub entry_point_name: Option<String>,
    #[serde(default)]
    pub manifest_sha256: Option<String>,
    pub manifest: Map<String, Value>,
    pub permissions_granted: Map<String, Value>,
    pub installed_at: DateTime<Utc>,
    #[serde(default)]
    pub enabled: bool,
    #[serde(default)]
    pub state: PluginActivationState,
    #[serde(default)]
    pub approved_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub approved_non_interactive: bool,
    #[serde(default)]
    pub last_seen_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub last_error: Option<String>,
    #[serde(default)]
    pub provenance: Option<Map<String, Value>>,
    #[serde(default)]
    pub legacy_source_url: Option<String>,
    #[serde(default)]
    pub legacy_source_ref: Option<String>,
    #[serde(default)]
    pub database_revision: Option<i64>,
    #[serde(default)]
    pub database_status: Option<String>,
}

impl PluginActivation {
    /// Non-legacy rows carry full distribution identity.
    pub fn validate(&self) -> Result<()> {
        if self.state == PluginActivationState::Legacy {
            return Ok(());
        }
        let mut missing = Vec::new();
        if self.distribution_name.is_none() {
            missing.push("distribution_name");
        }
        if self.entry_point_name.is_none() {
            missing.push("entry_point_name");
        }
        if self.manifest_sha256.is_none() {
            missing.push("manifest_sha256");
        }
        if self.approved_at.is_none() {
            missing.push("approved_at");
        }
        if missing.is_empty() {
            Ok(())
        } else {
            Err(Error::Validation(format!(
                "non-legacy plugin activation requires {}",
                missing.join(", ")
            )))
        }
    }
}

/// The `BudgetExceeded` exception payload: spend, limit, and window.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BudgetExceeded {
    pub spent: f64,
    pub limit: f64,
    pub window_days: i64,
}

impl BudgetExceeded {
    pub fn message(&self) -> String {
        format!(
            "LLM budget exceeded: ${:.2} spent in the last {}d against a ${:.2} limit.",
            self.spent, self.window_days, self.limit
        )
    }
}
